src/algorithms/Neal2.cpp

```diff
@@ -65,34 +58,34 @@ void Neal2<Hierarchy, Hypers, Mixing>::sample_allocations() {
     // Loop over clusters
     for (size_t k = 0; k < n_clust; k++) {
       // Probability of being assigned to an already existing cluster
-      probas(k) = this->mixing.mass_existing_cluster(cardinalities[k], n - 1) *
-                  unique_values[k].like(datum)(0);
+      probas(k) = mixing->mass_existing_cluster(cardinalities[k], n - 1) *
+                  unique_values[k]->like(datum)(0);
       if (singleton == 1 && k == allocations[i]) {
         // Probability of being assigned to a newly generated cluster
-        probas(k) = this->mixing.mass_new_cluster(n_clust, n - 1) *
-                    unique_values[0].eval_marg(datum)(0);
+        probas(k) = mixing->mass_new_cluster(n_clust, n - 1) *
+                    unique_values[0]->eval_marg(datum)(0);
       }
       tot += probas(k);
     }
     if (singleton == 0) {
       // Further update with marginal component
-      probas(n_clust) = this->mixing.mass_new_cluster(n_clust, n - 1) *
-                        unique_values[0].eval_marg(datum)(0);
+      probas(n_clust) = mixing->mass_new_cluster(n_clust, n - 1) *
+                        unique_values[0]->eval_marg(datum)(0);
       tot += probas(n_clust);
     }
     // Normalize
     probas = probas / tot;
 
     // Draw a NEW value for datum allocation
-    unsigned int c_new = stan::math::categorical_rng(probas, this->rng) - 1;
+    unsigned int c_new = stan::math::categorical_rng(probas, rng) - 1;
```

Comment:
Sure, where would this wrapper function be? Maybe in utils.hpp?

Reply:
or maybe something like 'distributions.hpp', you can move there also some functions on the multivariate normal lpdf

Reply:
Turns out that the correct file was already there, pushed by you, it's just that it was not used in the library. Anyway, should we be able to set the starting rng seed? If so, how?

Reply:
Which functions are you referring to?

Reply:
the lpdf method of HierarchyNNW, when we say:

`double exp = 0.5 * (tau_chol_factor_eval * (datum - mu).transpose()).squaredNorm();`

I think it's worth having a function with signature `double multi_normal_prec_lpdf(VectorXd datum, VectorXd mean, MatrixXd prec_chol, double prec_det);`
or something similar, so that we can use it in other places as well!

Reply:
Done. Please have a thorough look to make sure it is fine, as we are currently unable to test it